pkg/controller/direct/resourcemanager/tagkey_controller.go

```diff
@@ -0,0 +1,252 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/clients/generated/apis/tags/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+)
+
+// AddTagKeyController creates a new controller and adds it to the Manager.
+// The Manager will set fields on the Controller and start it when the Manager is started.
+func AddTagKeyController(mgr manager.Manager, config *controller.Config) error {
+	gvk := krm.TagsTagKeyGVK
+
+	// TODO: Share gcp client (any value in doing so)?
+	ctx := context.TODO()
+	gcpClient, err := newGCPClient(ctx, config)
+	if err != nil {
+		return err
+	}
+	m := &tagKeyModel{gcpClient: gcpClient}
+	return directbase.Add(mgr, gvk, m)
+}
+
+type tagKeyModel struct {
+	*gcpClient
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &tagKeyModel{}
+
+type tagKeyAdapter struct {
+	resourceID string
+
+	desired *krm.TagsTagKey
+	actual  *pb.TagKey
+
+	*gcpClient
+	tagKeysClient *api.TagKeysClient
+}
+
+// adapter implements the Adapter interface.
+var _ directbase.Adapter = &tagKeyAdapter{}
+
+// AdapterForObject implements the Model interface.
+func (m *tagKeyModel) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	tagKeysClient, err := m.newTagKeysClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &krm.TagsTagKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	resourceID := ValueOf(obj.Spec.ResourceID)
+	// resourceID is server-generated, no fallback
+	// TODO: How do we do resource acquisition - maybe by shortname?
+	resourceID = strings.TrimPrefix(resourceID, "tagKeys/")
+
+	return &tagKeyAdapter{
+		resourceID:    resourceID,
+		desired:       obj,
+		gcpClient:     m.gcpClient,
+		tagKeysClient: tagKeysClient,
+	}, nil
+}
+
+// Find implements the Adapter interface.
+func (a *tagKeyAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetTagKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	tagKey, err := a.tagKeysClient.GetTagKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	a.actual = tagKey
+
+	return true, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *tagKeyAdapter) Delete(ctx context.Context) (bool, error) {
+	// Already deletd
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteTagKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+
+	op, err := a.tagKeysClient.DeleteTagKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("deleting tagKey: %w", err)
```

Comment:
nit: here and elsewhere where it makes sense, do we want to also include the `tagKey` name? Assuming it doesn't bubble up in the direct controller.
```suggestion
		return false, fmt.Errorf("deleting tagKey %s: %w", a.fullyQualifiedName(), err)
```